### DISCUSSION TITLE
V2 release

### DISCUSSION THREAD

**Discussion by danielrbradley** [pulumi/pulumi-azure-native#1834] [Announcements]
It’s been over a year since we released our first stable version of the azure-native provider. Therefore we’re planning to do a major version upgrade with some limited breaking changes to give an opportunity to make some strategic improvements.

Please do use this discussion thread to ask any questions before or after the relase.

## Update Default Versions

Throughout version 1 we’ve endeavoured to keep backward compatibility of all the resources at the top level of each module – often continuing to use older versions of Azure’s REST APIs, and in some cases preview versions. 

For example in the EventGrid module, the EventSubscription resource currently uses version `2020-06-01` of the Azure REST API and “Channel” uses `2021-10-15-preview` when there is now `2022-06-15` available for both. Therefore we will update the versions we present to reflect the latest available Azure REST API versions. This will include some changes to resources which are not compatible between the two versions, for example a property might be renamed or its type changed. 

If you experience problems with incompatible types when upgrading to version 2 you can either:
- Update your code to align with the new resource’s arguments
- Continue to use the older version of the resource by updating your imports

To continue to use the same resource version as from version 1 of the provider package you will need to update your import statements as follows:

Check which “API Version” of the resource you are using in V1:

1. Read the inline code documentation for the resource
2. Find the resource in the [Azure Native registry API Docs](https://www.pulumi.com/registry/packages/azure-native/api-docs/)
3. Refer to the [v1.json version file](https://github.com/pulumi/pulumi-azure-native/blob/master/versions/v1.json)

Next, update your import to point to the specific older version. For example, to import the explicit version of the EventGrid EventSubscription resource:

### C#

```diff
- using EventGrid = Pulumi.AzureNative.EventGrid;
+ using EventGrid20200601 = Pulumi.AzureNative.EventGrid.V20200601;
```

### Typescript

```diff
- import { EventSubscription } from "@pulumi/azure-native/eventgrid";
+ import { EventSubscription } from "@pulumi/azure-native/eventgrid/v20200601";
```

### Python

```diff
- from pulumi_azure_native import eventgrid
+ from pulumi_azure_native.eventgrid import v20200601 as eventgrid20200601
```

### Go

```diff
- import eventgrid "github.com/pulumi/pulumi-azure-native/sdk/go/azure/eventgrid"
+ import eventgrid20200601 "github.com/pulumi/pulumi-azure-native/sdk/go/azure/eventgrid/v20200601”
```

### YAML

```diff
- type: azure-native:eventgrid:Topic
+ type: azure-native:eventgrid/v20200601:Topic
```

## Remove Deprecated Versions

Since v1.62.0 we have introduced deprecation notices on specific older versions of Azure’s REST API. At the point of releasing version 2 of the provider any deprecated versions will be removed. We recommend migrating any affected resources to a newer version of Azure’s REST API. 

Our estimates are that very few users should be affected by these changes as they only include versions which were released before 2021, have been superseded, and are not currently used as a default version.

The reason for these removals is to limit the growth rate of the SDKs. Most languages have limits on the size of an SDK; and we’ve reached a few already which have required some work to mitigate. We will continue to look at long term solutions to address these limitations but this is necessary work to ensure the stability of the provider.

## Simplify User Assigned Identities (work in progress)

User assigned identities are a property on many different resources and are currently quite tricky to pass correctly due to their type being unclear. As part of the move to version 2 of the provider, these user-assigned identities types will become simple lists of strings to make them much easier to use.

**Comment by JakeGinnivan** [comment]
These changes sound great, the compilation time of TypeScript is pretty bad right now. When we had a look at picking up Pulumi over Terraform (for Azure) and the editor performance and compilation time were a blocker.

One of the issues is deep imports still reference back to root modules so even if you try to pull in a subset of the types, TypeScript still needs to parse ~2million lines of types.

Is the plan to not re-export the older versions except from the deep imports?

**Reply by daniel-noland** [reply]
We have run into the same issue.

I switched from Webstorm to VS Code which helped a bit but it is still a very very heavy API.

I'm not 100% sure how much Pulumi can do about it if the API really just is gigantic.  But as you say @JakeGinnivan, if there was a way to import more modestly then I think that would help a lot.
